# **`Ejercicios 5: Funciones`**

### Ejercicio 1.7.1.

#### Crea una función que se llame escribir y que imprima por pantalla el texto “Estoy en la Escuela de Talento Digital”.

In [3]:
def escribir():
    print('Estoy en la Escuela de Talento Digital')

escribir()    

Estoy en la Escuela de Talento Digital


### Ejercicio 1.7.2.

#### Crea una función a la que le pases como parámetro un nombre e imprima por pantalla “nombre está en la Escuela de Talento Digital”.

In [8]:
def imprimir_presentacion(nombre):
    print(f'{nombre} está en la Escuela de Talento Digital')

imprimir_presentacion('Nahuel')
imprimir_presentacion('Elon Musk')
imprimir_presentacion('Sam Altman')

Nahuel está en la Escuela de Talento Digital
Elon Musk está en la Escuela de Talento Digital
Sam Altman está en la Escuela de Talento Digital


### Ejercicio 1.7.3.

#### Crea una función a la que le pases como parámetro una cifra e imprima en pantalla el valor de esa cifra multiplicado por tres.

In [11]:
def triplicar_numero(numero = 0):
    numero_triplicado = numero * 3
    print(f'{numero_triplicado}')

triplicar_numero()
triplicar_numero(1)
triplicar_numero(4)
triplicar_numero(7)

0
3
12
21


### Ejercicio 1.7.4.

#### Utilizando la función que has creado en el ejercicio 1.7.3., muestra por pantalla la tabla de multiplicar del 3. Por ejemplo, la tabla de multiplicar del 1 se mostraría así:

0 

1 

2 

3 

4 

5 

6 

7 

8 

9 

10 

In [13]:
def triplicar_numero(numero = 0):
    numero_triplicado = numero * 3
    print(f'{numero_triplicado}')

for num in range(11):
    triplicar_numero(num)
    print()


0

3

6

9

12

15

18

21

24

27

30



### Ejercicio 1.7.5.

#### Crea una función a la que le pases dos números e imprima por pantalla la suma de ambos.

In [14]:
def suma_numeros(a=0, b=0):
    suma = a + b
    print(suma)
    
suma_numeros()
suma_numeros(1,5)
suma_numeros(20,100)

0
6
120


### Ejercicio 1.7.6.

#### Crea una función con un parámetro que muestre por pantalla “El número es par” si el parámetro introducido es par y que muestre “El número es impar” si el parámetro es impar.

In [15]:
def es_par(num):
    num_par = num % 2 == 0
    if num_par:
        print('El número es par')
    else:
        print('El número es impar')

es_par(0)
es_par(4)
es_par(5)
es_par(1001)
es_par(1040)

El número es par
El número es par
El número es impar
El número es impar
El número es par


### Ejercicio 1.7.7.

#### Crea una función que reciba un número y muestre por pantalla su factorial.

In [30]:
def factorial(num):
    res = 1
    if num > 0:
        while num > 0:
            res *= num
            num -= 1
        print(res)
    elif num == 0:
        print(1)
    else:
        print('Math ERROR')

factorial(4)
factorial(6)
factorial(8)
factorial(0)
factorial(-4)

24
720
40320
1
Math ERROR


### Ejercicio 1.7.8.

#### Crea una función que calcule la media de los números que pasemos como parámetros y la muestre por pantalla. Ten en cuenta que podemos pasarle dos números como podemos pasarle diez, no lo tenemos definido en un principio.

In [46]:
def calcular_media(*nums):
    suma = sum(nums)
    cantidad = len(nums)
    media = suma/cantidad
    print(round(media,2))

calcular_media(8,9,10)
calcular_media(8,9,9,10)
calcular_media(5,7,9,10)

9.0
9.0
7.75


### Ejercicio 1.7.9.

#### Crea una función que dados dos números muestre por pantalla todos los números que hay entre ellos, incluidos ambos.

In [10]:
def imprimir_numeros(a:int, b:int):
    for num in range(a, b+1):
        print(num, end=' ')
    print()    

imprimir_numeros(-4,11)
imprimir_numeros(15,40)

-4 -3 -2 -1 0 1 2 3 4 5 6 7 8 9 10 11 
15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 


### Ejercicio 1.7.10.

#### Crea una función que dados una serie de números identifique el máximo y el mínimo y muestre por pantalla “El máximo es: x”, siendo x el mayor de los números, y “El mínimo es: y”, siendo y el menor de los números.

In [13]:
def imprimir_max_min(*nums):
    maximo = max(nums)
    minimo = min(nums)
    print(f'El máximo es: {maximo}')
    print(f'El mínimo es: {minimo}')

imprimir_max_min(2,3,46,1,33,0,100)    

El máximo es: 100
El mínimo es: 0


### Ejercicio 1.7.11.

#### Crea una función a la que se introduzca un texto como parámetro y muestre por pantalla el número de caracteres que tiene ese texto.

In [19]:
def imprime_nro_caracteres(texto:str):
    cantidad = 0
    for caracter in texto:
        cantidad += 1
    print(f'El número de caracteres que tiene "{texto}" es de {cantidad}')    

def imprime_longitud(texto:str):
    caracteres = len(texto)
    print(f'El número de caracteres que tiene "{texto}" es de {caracteres}')
    
imprime_nro_caracteres('hola mundo') 
imprime_longitud('Soy un Python Data Analyst')

El número de caracteres que tiene "hola mundo" es de 10
El número de caracteres que tiene "Soy un Python Data Analyst" es de 26


### Ejercicio 1.7.12.

#### Crea un programa que te de la opción de cambiar horas, minutos y segundos a segundos o segundos a horas, minutos y segundos. Para ello sigue los siguientes pasos:

Crea una función a la que le introduzcas tres parámetros (horas, minutos, segundos) y que calcule a cuántos segundos equivalen, mostrando el resultado por pantalla.
Crea una función a la que le introduzcas un parámetro (segundos) y muestre por pantalla a cuantas horas, minutos y segundos corresponderían a esos segundos.
Crea un programa que te muestre dos opciones:
                    
      (1) Convertir a segundos

      (2) Convertir a horas, minutos y segundos

      Si el usuario elige la opción 1, se le pedirán las horas, minutos y segundos que quiere transformar en segundos y se usará la primera función para mostrar por pantalla el número de segundos correspondiente.

      Si el usuario elige la opción 2, se le pedirán los segundos, se convertirán en horas, minutos y segundos usando la segunda función para mostrar el resultado por pantalla.

In [31]:
def tiempo_segundos(horas:int, minutos:int, segundos:int):
    tiempo = segundos + minutos * 60 + horas * 3600
    print(f'El tiempo en segundos es: {tiempo} segundo/s')

def tiempo_hms(segundos):
    horas = segundos // 3600 # calculando las horas
    minutos = (segundos % 3600) // 60 # calculando los minutos del resto del tiempo de horas
    segundos_restantes = segundos % 60 # calculando los segundos del resto de la división entre cantidad de segundos y 60
    print(f'{segundos} segundos son: {horas} hora/s, {minutos} minuto/s y {segundos_restantes} segundo/s')
    
while True:
    print('¿Qué te gustaría hacer?')
    opcion = int(input('1. Convertir el tiempo a segundos \n2. Convertir los segundos a horas, minutos y segundos \n'))
    if opcion == 1:
        horas = int(input('Introduce las horas:'))
        minutos = int(input('Introduce los minutos:'))
        segundos = int(input('Introduce los segundos:'))
        tiempo_segundos(horas, minutos, segundos)
        break
    elif opcion == 2:
        segundos = int(input('Introduce los segundos:'))
        tiempo_hms(segundos)
        break
    else:
        print('Ingresa el número 1 o 2 para las opciones.')

¿Qué te gustaría hacer?


1. Convertir el tiempo a segundos 
2. Convertir los segundos a horas, minutos y segundos 
 2
Introduce los segundos: 7435


7435 segundos son: 2 hora/s, 3 minuto/s y 55 segundo/s


### Ejercicio 1.7.13.

#### Crea un programa que le pida al usuario texto y compruebe si este contiene la @. En caso de que la contenga devolverá “El texto contiene la @” y en caso contrario “El texto no contiene la @”. Utiliza una función para realizar la comprobación.

In [2]:
def comprueba_arroba(texto:str):
    if '@' in texto:
        print(f'El texto "{texto}" contiene la @')
    else:
        print(f'El texto "{texto}" no contiene la @')

comprueba_arroba('hola mundo')
comprueba_arroba('nahuel.developer1@gmail.com')

El texto "hola mundo" no contiene la @
El texto "nahuel.developer1@gmail.com" contiene la @


### *Ejercicio 1.7.14.*

#### Crea un programa que pida al usuario una cantidad y un porcentaje de impuestos a pagar y calcule el importe total de la cantidad más el porcentaje de impuestos indicado y lo muestre por pantalla. Es decir, si mi factura es de 300 y tengo que sumarle el 10% de impuestos, el importe total sería 330. 

#### Utiliza una función con dos parámetros, el importe inicial y el porcentaje de impuestos, y que devuelva el importe total.

In [10]:
def calcula_importe(cantidad:float, impuesto:float): 
    total = cantidad * (1 + impuesto / 100)
    print('El importe total a pagar es:', round(total,2))

cantidad = float(input('Hola, ingrese una cantidad a pagar:'))
impuesto = float(input('Ingrese el porcentaje de impuesto correspondiente:'))
calcula_importe(cantidad, impuesto)

Hola, ingrese una cantidad a pagar: 1000
Ingrese el porcentaje de impuesto correspondiente: 35


El importe total a pagar es: 1350.0


### Ejercicio 1.7.15.

#### Crea una función en la que el usuario pueda introducir todos los números que quiera, siendo el último el cero, y calcule la suma de todos ellos. Después ejecútala tres veces con tres grupos de números diferentes y muestra por pantalla el máximo resultado siguiendo a la frase “El máximo es x".

In [13]:
def suma_numeros():
    numero = float(input('Ingrese un número. Si no desea seguir ingresando números, escriba 0 (cero):'))
    contador = 0
    while numero != 0:
        contador += numero
        numero = float(input('Ingrese un número. Si no desea seguir ingresando números, escriba 0 (cero):'))
    return contador

print('Escriba la primera tanda de números:')
suma1 = suma_numeros()
print('Escriba la segunda tanda de números:')
suma2 = suma_numeros()
print('Escriba la tercera tanda de números:')
suma3 = suma_numeros()

print(f'El máximo es:', max(suma1, suma2, suma3))


Escriba la primera tanda de números:


Ingrese un número. Si no desea seguir ingresando números, escriba 0 (cero): 2
Ingrese un número. Si no desea seguir ingresando números, escriba 0 (cero): 3
Ingrese un número. Si no desea seguir ingresando números, escriba 0 (cero): 0


Escriba la segunda tanda de números:


Ingrese un número. Si no desea seguir ingresando números, escriba 0 (cero): 1
Ingrese un número. Si no desea seguir ingresando números, escriba 0 (cero): 2
Ingrese un número. Si no desea seguir ingresando números, escriba 0 (cero): 3
Ingrese un número. Si no desea seguir ingresando números, escriba 0 (cero): 0


Escriba la tercera tanda de números:


Ingrese un número. Si no desea seguir ingresando números, escriba 0 (cero): 1
Ingrese un número. Si no desea seguir ingresando números, escriba 0 (cero): 2
Ingrese un número. Si no desea seguir ingresando números, escriba 0 (cero): 3
Ingrese un número. Si no desea seguir ingresando números, escriba 0 (cero): 4
Ingrese un número. Si no desea seguir ingresando números, escriba 0 (cero): 5
Ingrese un número. Si no desea seguir ingresando números, escriba 0 (cero): 6
Ingrese un número. Si no desea seguir ingresando números, escriba 0 (cero): 0


El máximo es: 21.0


Si tienes el siguiente código:

def maximo(a,b): 

   if x>y: 
       return x 
   else: 
       return y 

def minimo(a,b): 

   if x<y: 
       return x 
   else: 
       return y 

x=int(input("Un número: ")) 

y=int(input("Otro número: ")) 

print(maximo(x-3, minimo(x+2, y-5)))

Si tienes los valores x = 5 e y = 1 y lo ejecutas, por pantalla mostrará que la solución es 5, pero, realmente, debería mostrar que la solución es 2. El código es erróneo, pero ¿cuál es el error? 

print(maximo(5-3, minimo(5+2, 1-5)))
print(maximo(2, minimo(7, -4)))

In [18]:
def maximo(a,b): 
    if x > y: 
        return x 
    else: 
        return y 

def minimo(a,b):
    if x<y: 
        return x 
    else: 
        return y 
x = 5
y = 1
# print(maximo(x-3, minimo(x+2, y-5)))
print(maximo(5-3, minimo(5+2, 1-5)))
# Notar que esta función no hace uso de los argumentos, sino de los valores x e y únicamente

5


In [19]:
# ejercicio corregido
def maximo(x,y): 
    if x > y: 
        return x 
    else: 
        return y 

def minimo(x,y):
    if x<y: 
        return x 
    else: 
        return y 
x = 5
y = 1

print(maximo(x-3, minimo(x+2, y-5)))

2


### Ejercicio 1.7.17.

#### En el ejercicio 1.7.7, has creado una función que calculaba el factorial de un número, pero sin utilizar recursividad. Crea ahora una función que calcule el factorial de un número, utilizando recursividad.

In [25]:
def factorial(num):
    if num > 0:
        return num * factorial(num - 1)
    elif num == 0:
        return 1
    else:
        return 'Math ERROR'

print(factorial(4))
print(factorial(6))
print(factorial(8))
print(factorial(0))
print(factorial(-4))

24
720
40320
1
Math ERROR


### Ejercicio 1.7.18.

#### Utilizando la función creada en el ejercicio anterior, crea un programa que solicite números al usuario y, mientras este no incluya el número -1, le devuelva el factorial del número introducido y le solicite otro número. Cuando el usuario incluya el número -1, el programa finalizará y le mostrará al usuario de cuántos números ha calculado el factorial. Un ejemplo de ejecución sería:

#### Incluye un número (si incluyes el -1 el programa se detiene): 6 

#### El factorial de 6 es 720 

#### Incluye un número (si incluyes el -1 el programa se detiene): 4 

#### El factorial de 4 es 24 

#### Incluye un número (si incluyes el -1 el programa se detiene): -1 

#### Has calculado el factorial de 2 números. 

In [8]:
def factorial(num):
    if num > 0:
        return num * factorial(num - 1)
    elif num == 0:
        return 1
    else:
        return 'Math ERROR'

contador = 0    
while True:
    nro_usuario = int(input('Ingrese un número, si ingresa -1 el programa finalizará:'))
    if nro_usuario == -1:
        print(f'Has calculado el factorial de {contador} número/s. Fin del programa')
        break
    contador += 1    
    print(f'El factorial de {nro_usuario} es {factorial(nro_usuario)}')       

Ingrese un número, si ingresa -1 el programa finalizará: 0


El factorial de 0 es 1


Ingrese un número, si ingresa -1 el programa finalizará: -3


El factorial de -3 es Math ERROR


Ingrese un número, si ingresa -1 el programa finalizará: 4


El factorial de 4 es 24


Ingrese un número, si ingresa -1 el programa finalizará: 5


El factorial de 5 es 120


Ingrese un número, si ingresa -1 el programa finalizará: -1


Has calculado el factorial de 4 número/s. Fin del programa


### Ejercicio 1.7.19

#### Crea una función que, dado un número, calcule la suma de sus dígitos. 

In [14]:
def suma_digitos(numero):
    suma = 0
    for digito in str(numero):
        suma += int(digito)
    print(suma)    

suma_digitos(123)      
suma_digitos(123456)

6
21


### Ejercicio 1.7.20.

#### Crea una función que, dado un número, compruebe si un número es primo. En caso de que lo sea, la función devolverá True, en caso de que no lo sea, devolverá False.

In [20]:
# una forma de hacerla, recorriendo hasta el número anterior al numero y comprobando uno a uno los restos.
def es_primo(numero):
    if numero < 2:
        return False
    for i in range(2, numero):
        if numero % i == 0: # con hallar un divisor distinto de 1 y el mismo número, ya no es primo
            return False
    return True

# Otra forma de hacerlo, pero más eficiente
# def es_primo(numero):
#     if numero < 2:
#         return False
#     for i in range(2, int(numero**0.5)+1):
#         if numero % i == 0:
#             return False
#     return True

print(es_primo(2))
print(es_primo(16))
print(es_primo(17))
print(es_primo(22))
print(es_primo(23))

True
False
True
False
True


### Ejercicio 1.7.21.

#### Utilizando las funciones que has creado en los ejercicios 1.7.17, 1.7.19 y 1.7.20, crea un programa que solicite al usuario que introduzca un número, compruebe si es primo, en ese caso, que calcule la suma de sus dígitos y, en caso contrario, que devuelva su factorial.

In [31]:
def factorial(num):
    if num > 0:
        return num * factorial(num - 1)
    elif num == 0:
        return 1
    else:
        return 'Math ERROR'

def suma_digitos(numero):
    suma = 0
    for digito in str(numero):
        suma += int(digito)
    return suma    

def es_primo(numero):
    if numero < 2:
        return False
    for i in range(2, numero):
        if numero % i == 0:
            return False
    return True

nro_usuario = int(input('Ingrese un número entero:'))

if es_primo(nro_usuario):
    print(f'La suma de los digitos de {nro_usuario} es: {suma_digitos(nro_usuario)}')
else:
    print(f'El factorial de {nro_usuario} es: {factorial(nro_usuario)}')
        

Ingrese un número entero: 41


La suma de los digitos de 41 es: 5


### Ejercicio 1.7.22.

#### Crea un programa que muestre por pantalla cuántos días tiene el mes del año concreto que te indique un usuario. Para ello necesitas crear dos funciones, una de ellas que te devuelva True si el año introducido es bisiesto, y otra que indique el número de días que tiene un mes de un año concreto, ambos introducidos como parámetros.

#### Prueba con dos meses y años diferentes

In [36]:
def es_bisiesto(anio):
    # Devuelve True si el año introducido es bisiesto, False en caso contrario.
    # Cualquier año divisible por 4 es un año bisiesto
    return anio % 4 == 0 and (anio % 100 != 0 or anio % 400 == 0)

def dias_del_mes(mes, anio):
    # Devuelve el número de días que tiene un mes de un año concreto.
    if mes == 2:
        return 29 if es_bisiesto(anio) else 28
    elif mes in [4, 6, 9, 11]:
        return 30
    else:
        return 31

# Pedimos al usuario que introduzca el mes y el año
mes = int(input("Introduce el mes (1-12): "))
anio = int(input("Introduce el año: "))

# Mostramos por pantalla el número de días que tiene el año con el mes introducido
print(f"El mes {mes} del año {anio} tiene {dias_del_mes(mes, anio)} días.")


Introduce el mes (1-12):  4
Introduce el año:  2023


El mes 4 del año 2023 tiene 30 días.


### Ejercicio 1.8.1.

#### Hay una librería preinstalada en Python que permite trabajar con fechas y horas. Impórtala y, con uno de los objetos definidos en ella, que te permite representar fechas, utiliza una función ya definida para mostrar por pantalla la fecha de hoy. 

In [25]:
from datetime import datetime as time

fecha_1 = time.today()
fecha_2 = date.today()
print(fecha_1)
print(fecha_2)

2023-04-13 00:55:59.237963
2023-04-13


### Ejercicio 1.8.2.

#### Ahora, importa solo el objeto que te permite representar fechas y, utilizando la misma función del ejercicio anterior, muestra por pantalla la fecha de hoy.

In [19]:
from datetime import date

fecha_actual = date.today()
print(fecha_actual)

2023-04-13


### Ejercicio 1.8.3.

#### Utilizando el mismo objeto del ejercicio anterior, data la fecha (2023, 11, 25) y sustitúyela por la fecha (2023, 12, 26) utilizando el método replace, ya definido en ese objeto.

In [23]:
from datetime import date

fecha = date(2023, 11, 25)
fecha_corregida = fecha.replace(2023, 12, 26)
print(fecha)
print(fecha_corregida)

2023-11-25
2023-12-26


### Ejercicio 1.8.4.

#### ¿Qué función de la librería random utilizarías si quieres generar un número decimal aleatorio entre 0.0 y 1.0?

In [93]:
import random as ran

def generar_nro_aleatorio():
    nro = round(ran.random(), 4)
    print(nro)
    
generar_nro_aleatorio()
generar_nro_aleatorio()
generar_nro_aleatorio()
generar_nro_aleatorio()

0.5373
0.2948
0.5766
0.9424


### Ejercicio 1.8.5.

#### Y si lo que quieres es generar un número decimal aleatorio entre 4 y 6, ¿qué función de la librería random utilizarías? Y si el número no es decimal, sino entero, ¿utilizarías una función diferente?


In [95]:
import random as ran

# para número decimal aleatorio entre 4 y 6
nro_decimal_aleatorio = ran.uniform(4,6)
print(nro_decimal_aleatorio)

# para número entero aleatorio entre 4 y 6
nro_entero_aleatorio = ran.randint(4,6)
print(nro_entero_aleatorio)

5.247096499984588
4


### Ejercicio 1.8.6.

#### ¿Qué librería preinstalada en Python te permite trabajar con archivos ZIP?

#### *`zipfile`*

Tomado de la documentación oficial de Python, Python.org, zipfile permite trabajar con archivos ZIP. 
"Este módulo proporciona herramientas para crear, leer, escribir, agregar y listar un archivo ZIP".

### Ejercicio 1.8.7.

#### ¿Qué librería preinstalada en Python te permite leer y escribir archivos csv?

#### *`CSV`*

La librería CSV (Comma Separated Values) permite leer y escribir archivos csv